accepted/0000-yomi-initial-integration.md

```diff
@@ -0,0 +1,361 @@
+consequences, as the state detect early that the root filesystem
+device is already mounted.
+
+Some minor bugs in Uyuni (bsc#1150586, bsc#1150589, bsc#1150590,
+bsc#1152673, bsc#1152767) makes this process less direct, but for all
+except bsc#1152673 Yomi is providing a workaround. The remaining bug
+requires some manual intervention before the highstate can be applied
+from Uyuni.
+
+The big problem for this workflow has been pointed by Joachim Werner,
+and is about the default values for the pillar data. IIUC the problem
+is that if Yomi provides some default pillar configuration, the new
+discovered node can wrongly apply a highstate (that AFAIK are applied
+automatically or scheduled by the user) with non-optimal data (disk
+configuration, packages, repositories, etc)
+
+There are two proposals, also from Joe, to address this issue (copied
+verbatim from the original post):
+
+-->-->--
+
+1) Discover first.
+
+- Boot a new node (via PXE, UEFI network boot, ISO)
+
+- Register it in SUSE Manager and populate the hardware data
+
+- Shut it down (or leave it running in "wait mode")
+
+- Apply custom configuration by adding formulas and filling in forms
+  (could also be "meta-templated" by SUSE Manager's "activation keys"
+  that define, channels, admin users, groups)
+
+- Boot if necessary
+
+- Apply highstate, including yomi state
+
+2) Pre-populate the DB:
+
+- Create a new "empty" node in the DB and provide the data needed to
+  match it with a physical system (e.g. MAC address or serial number)
+
+- Apply custom configuration by adding formulas and filling in forms
+  (could also be "meta-templated" by SUSE Manager's "activation keys"
+  that define, channels, admin users, groups)
+
+- Boot node (via PXE, UEFI network boot, ISO)
+
+- Register it in SUSE Manager and populate the hardware data
+
+- Apply highstate, including yomi state
+
+(1) exists as a feature, but currently the "discovery image" isn't
+using Salt yet. We could re-use the exact same image that you're using
+for installation, just stop before applying the yomi state.
+
+(2) also exists for saltboot. But as of today it's just an API call,
+no UI version yet. Trivial to add. What's missing is applying an
+activation key to such a system as a template. As of today you'd have
+to inject the activation key grain on the minion side before
+registering the system to Manager.
+
+--<--<--
+
+## Disabling automatic highstate
+
+If I understand the issue correctly, the fact that the Yomi pillar is
+split into multiple sub-forms makes the application of partially
+defined pillar data more probable that if it is defined in a single,
+big form, without defaults.
+
+To potentially fix this issue, I propose in this RFC to have a grain
+emitted by the just-discovered system that will disable any automatic
+highstate application, if present. Only the highstate for this system
+can be manually applied:
+
+```yaml
+  automatic-highstate: no
+```
```

Comment:
I don't particularly like this part. Actually, I'd _expect_ the complete high state to be applied during installation (if that's possible in one go without reboot). The problem isn't _partially_ filled in pillar data. It's that without magic a newly added system has _no_ states and pillars assigned, unless you add global defaults somewhere (e.g. in a manually edited top.sls). This can be fixed by either pre-creating the exact system in the database and assigning all the required pillar data via Formulas with Forms (see above) and software channels (not possible yet?) or preparing defaults via system groups and then using an activation code to assign systems to their "templates". Let's double-check with the retail team how they are doing it. I guess they have hardware-specific groups for the defaults and do some magic with a reactor at install time to add the system to the right group, but I haven't checked yet.

Reply:
> I don't particularly like this part. Actually, I'd _expect_ the complete high state to be applied during installation (if that's possible in one go without reboot). 

Actually I expect the same. I want the high state be applied during the installation. This is about a different issue: I do not want SUMA to schedule and automatic highstate for this node that I booted via PXE boot the Yomi image, and I want to install from scratch.

Also a full highstate have other consequences that are missing in the RFC (I will add those later): the main problem is that Yomi is talking about the to-be-chrooted environment, and the other states are talking about the system in place (that first is the Yomi image, and later will be the already installed system).

For example, if we take a new node, boot it via the Yomi image, register the salt-minion key in the manager, and register the bind and the yomi formulas, the bind one is talking about the Yomi image (the system that is in the memory), meanwhile the yomi formula is talking about the system that will be chrooted.

> The problem isn't _partially_ filled in pillar data. It's that without magic a newly added system has _no_ states and pillars assigned, unless you add global defaults somewhere (e.g. in a manually edited top.sls).

This confuses me. When I boot a system from the Yomi image, and I accept the minion key, I am able to assign some formulas from the catalog.

Not having any state and pillars pre-assigned is good (IMHO), as give the opportunity to manually assign the yomi formulas to bootstrap the installation.

But if the new system have some state and some pillar pre-assigned and one of them is Yomi with some generic (and incorrect) pillar data, this can be a problem.

In the default case the pillar data after the discovery is empty, so if meanwhile a highstate is automatically applied, I expect that the Yomi states will fail (but nothing bad will happen)

As meanwhile I have my pillar data in multiple forms, I also was expecting that meanwhile I am adding the next subform, an automatic highstate is applied in my system, maybe causing an undesired reboot.

This is why I propose a new grain: to avoid any automatic state.highstate.

If this mental model is not correct, I need to fix this ASAP.

> This can be fixed by either pre-creating the exact system in the database and assigning all the required pillar data via Formulas with Forms (see above) and software channels (not possible yet?) or preparing defaults via system groups and then using an activation code to assign systems to their "templates". Let's double-check with the retail team how they are doing it. I guess they have hardware-specific groups for the defaults and do some magic with a reactor at install time to add the system to the right group, but I haven't checked yet.

What is the difference between starting with no state and no pillar, assigning yomi from the formula catalog and filling the pillars, from pre-assigning the yomi formula and the adapted pillar data and later registering the system?

Sorry if I am a bit obtuse in this point, I am clearly missing some data.

Reply:
As for not triggering another automated highstate: Those are all triggered from Java based on registration events if I understand things correctly. We can add logic there to determine whether we are in the installation phase or not.

As for assigning formulas to a to-be-deployed system: The thing is, that you may want to assign all configuration _before_ you install the system, not boot/register, then manually add formulas to the already registered system.

Of course you can _still_ just generically boot a node and then assign it a configuration. That's actually _exactly_ my first scenario, where the minion will automatically stay in a "wait" state, with the Salt minion running from the RAM disk, as long as the installation highstate isn't triggered. I just think that the more common case would be to plan things ahead (including host names, roles etc.) and pre-create those systems in the database. OR use an activation key grain. That way you can define the complete "template" server-side, but still allow injecting something on the client side that lets you choose the role.

Don't worry, we don't _usually_ apply automated regular. high states anywhere, only have that as an option during the initial onboarding. You don't need a grain to suppress it.

Some customers want scheduled highstates like in Puppet, but once we implement that, we can always deactivate it until the installation phase is over (or, as mentioned in my notes, introduce a separate state for installation).

The difference between starting with no state/pillar and starting with a pre-assigned configuration is just two different ways how customers work. We've had the same two patterns in SLEPOS as well: Some customers want to have all their systems pre-defined before the rollout (e.g. every single cash register, with host name defined by exact position in the store, matching to an exact MAC address). Others will want to just bring in a new box and determine last minute (e.g. with an activation key) what role it should get.

We can easily support both, as we already can create "empty" systems (at least via the API).

Reply:
> As for assigning formulas to a to-be-deployed system: The thing is, that you may want to assign all configuration _before_ you install the system, not boot/register, then manually add formulas to the already registered system.

I understand. We want to define the full highstate (yomi and what is not yomi), and apply it in a single step during the provisioning. 

To achieve this we need to make the states that are not part of yomi aware of the chroot when this is present, and unaware of it when it is not present.

I added some new comments about this in the RFC, under the "Unresolved questions" section (Is Yomi part of the high state?). This is a big topic that was missing.